pipeline/metadata/satellite.py

```diff
@@ -396,14 +396,60 @@ def _unflatten_satellite(flattened_measurement: Iterable[Row]) -> Iterator[Row]:
         combined['received'] += received
       elif received:
         combined['received'].append(received)
-    combined.pop('measurement_id')
     # Remove extra tag fields from the measurement. These may be added
     # during the tagging step if a vantage point also appears as a response IP.
     for tag in ['asname', 'asnum', 'http', 'cert']:
       combined.pop(tag, None)
+
+    combined.pop('roundtrip_id')  # Remove roundtrip id
     yield combined
 
 
+def flatten_received_ips(dns_roundtrip: Row) -> Iterator[Row]:
+  """Flatten a row with multiple received ips into rows with a single ip.
+
+  Args:
+    row element like
+      {
+        "field": "value"
+        "received" [{
+            'ip': '1.2.3.4'
+        },{
+            'ip': '5.6.7.8'
+        }]
+      }
+
+  Returns:
+    Rows like
+      {
+        "field": "value"
+        "roundtrip_id" = "a"
+        "received" [{
+            'ip': '1.2.3.4'
+        }]
+      }
+      {
+        "field": "value"
+        "roundtrip_id" = "a"
+        "received" [{
+            'ip': '5.6.7.8'
+        }]
+      }
+  """
+  # This id is used to reconstruct the row structure when unflattening
+  dns_roundtrip['roundtrip_id'] = uuid.uuid4().hex
+
+  all_received = dns_roundtrip['received']
+
+  if len(all_received) == 0:
+    yield dns_roundtrip
+    return
+
+  for received in all_received:
+    dns_roundtrip['received'] = [received]
```

Comment:
This is type abuse. The new field is no longer a list, and the structure has a different semantics. You need to define a new type. But see my refactoring suggestion above. By not carrying the entire observation around, you are forced to not reuse the type.